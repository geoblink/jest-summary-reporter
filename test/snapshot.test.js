const {execFile} = require('child_process');

/**
 * Run jest using a given configfile and the args, and return the
 * stderr (all reporters go to stderr).
 * 
 * The configfile should be specified relative to the fixtures 
 * directory, so just the filename will suffice.
 * 
 * This assumes it is being run from the top-level directory of
 * this project and hardcodes a path to jest.
 */
function runJest(configFile, ...args) {
    return new Promise(resolve => {
        execFile('node_modules/.bin/jest', ['-c', "test/fixture/"+configFile, ...args], (err, stdout, stderr) => {
            let code = err ? err.code : 0;
            stderr = fuzzOutput(stderr);
            resolve({code, stdout, stderr});
        });
    });
}

/**
 * Alter the outout to make it stable from run to run.
 * We remove all outoput from the jest default reporters,
 * because the order of tests is impossible to force.
 * @param {String} output 
 */
function fuzzOutput(output) {
    return output.replace(/[\s\S]+Ran all test suites.*?$/m, '');
}

describe('with default configuration', () => {
    it('runs tests showing failures at end', async () => {
        const result = await runJest('jest-summary.config.json');
        expect(result).toMatchSnapshot();
    });    
    it('final message indicates when tests all pass', async () => {
        const result = await runJest('jest-summary.config.json', '-t', 'pass');
        expect(result).toMatchSnapshot();
    });    
});
describe('with failuresOnly=false configuration', () => {
    it('runs tests showing all tests at end', async () => {
        const result = await runJest('jest-summary-all.config.json');
        expect(result).toMatchSnapshot();
    });
});
describe('when misconfigured to be only reporter', () => {
    it('runs tests and returns code=1', async () => {
        const result = await runJest('jest-summary-only.config.json');
        expect(result).toMatchSnapshot();
        expect(result.code).toEqual(1);
    });
});
