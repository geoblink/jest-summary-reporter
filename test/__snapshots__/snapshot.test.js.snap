// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when misconfigured to be only reporter runs tests and returns code=1 1`] = `
Object {
  "code": 1,
  "stderr": "
Summary of failed tests

test/fixture/fail.test.js
  ✕ fails
  nested
    deeply nested
      ✕ fails
test/fixture/fail2.test.js
  ✕ fails
  nested
    deeply nested
      ✕ fails

",
  "stdout": "",
}
`;

exports[`with default configuration runs tests showing failures and their messages at end 1`] = `
Object {
  "code": 1,
  "stderr": "

Failure messages

test/fixture/fail.test.js
  ● fails

    expect(received).toEqual(expected) // deep equality

    Expected: 2
    Received: 1

      1 | it('fails', () => {
    > 2 |   expect(1).toEqual(2);
        |             ^
      3 | });
      4 | it('passes', () => {
      5 |   expect(2).toEqual(2);

      at Object.<anonymous> (test/fixture/fail.test.js:2:13)

  ● nested › deeply nested › fails

    expect(received).toEqual(expected) // deep equality

    Expected: 1
    Received: 2

      17 |     });
      18 |     it('fails', () => {
    > 19 |       expect(2).toEqual(1);
         |                 ^
      20 |     });
      21 |   });
      22 | });

      at Object.<anonymous> (test/fixture/fail.test.js:19:17)

test/fixture/fail2.test.js
  ● fails

    expect(received).toEqual(expected) // deep equality

    Expected: 2
    Received: 1

      1 | it('fails', () => {
    > 2 |   expect(1).toEqual(2);
        |             ^
      3 | });
      4 | it('passes', () => {
      5 |   expect(2).toEqual(2);

      at Object.<anonymous> (test/fixture/fail2.test.js:2:13)

  ● nested › deeply nested › fails

    expect(received).toEqual(expected) // deep equality

    Expected: 1
    Received: 2

      17 |     });
      18 |     it('fails', () => {
    > 19 |       expect(2).toEqual(1);
         |                 ^
      20 |     });
      21 |   });
      22 | });

      at Object.<anonymous> (test/fixture/fail2.test.js:19:17)


Summary of failed tests

test/fixture/fail.test.js
  ✕ fails
  nested
    deeply nested
      ✕ fails
test/fixture/fail2.test.js
  ✕ fails
  nested
    deeply nested
      ✕ fails

",
  "stdout": "",
}
`;

exports[`with failuresOnly=false and shouldPrintFailureMessages=false configuration runs tests showing all tests at end 1`] = `
Object {
  "code": 1,
  "stderr": "

Summary of tests

test/fixture/fail.test.js
  ✕ fails
  ✓ passes
  nested
    ✓ nested pass
    ○ skipped test
  nested
    deeply nested
      ✓ passes
      ✕ fails
test/fixture/fail2.test.js
  ✕ fails
  ✓ passes
  nested
    ✓ nested pass
    ○ skipped test
  nested
    deeply nested
      ✓ passes
      ✕ fails

",
  "stdout": "",
}
`;

exports[`with shouldPrintFailureMessages=false configuration final message indicates when tests all pass 1`] = `
Object {
  "code": 0,
  "stderr": "

All tests passed


",
  "stdout": "",
}
`;

exports[`with shouldPrintFailureMessages=false configuration runs tests showing failures at end 1`] = `
Object {
  "code": 1,
  "stderr": "

Summary of failed tests

test/fixture/fail.test.js
  ✕ fails
  nested
    deeply nested
      ✕ fails
test/fixture/fail2.test.js
  ✕ fails
  nested
    deeply nested
      ✕ fails

",
  "stdout": "",
}
`;
